src/main/java/com/amazon/ionpathextraction/PathExtractorImpl.java

```diff
@@ -72,48 +72,82 @@ public void match(final IonReader reader, final T context) {
 
         final Tracker<T> tracker = new Tracker<>(maxSearchPathDepth, searchPaths, reader.getDepth());
 
-        matchRecursive(reader, tracker, context);
+        matchAllValuesRecursive(reader, tracker, context);
     }
 
-    private int matchRecursive(final IonReader reader, final Tracker<T> tracker, final T context) {
-        final int currentDepth = tracker.getCurrentDepth();
-        int readerContainerIndex = 0;
+    @Override
+    public void matchCurrentValue(final IonReader reader) {
+        matchCurrentValue(reader, null);
+    }
 
-        while (reader.next() != null) {
-            // will continue to next depth
-            final List<SearchPath<T>> partialMatches = new ArrayList<>();
-
-            for (SearchPath<T> sp : tracker.activePaths()) {
-                final MatchContext matchContext = new MatchContext(reader, currentDepth, readerContainerIndex, config);
-                // a terminal search path is at the last path component meaning that if this search path partially
-                // matches it will be a full match and the callback must be invoked
-                boolean searchPathIsTerminal = isTerminal(tracker.getCurrentDepth(), sp);
-                boolean partialMatch = sp.partialMatchAt(matchContext);
-
-                if (partialMatch) {
-                    if (searchPathIsTerminal) {
-                        int stepOutTimes = invokeCallback(reader, sp, tracker.getInitialReaderDepth(), context);
-                        if (stepOutTimes > 0) {
-                            return stepOutTimes - 1;
-                        }
-                    } else {
-                        partialMatches.add(sp);
+    @Override
+    public void matchCurrentValue(final IonReader reader, final T context) {
+        checkArgument(reader.getDepth() == 0 || config.isMatchRelativePaths(),
+            "reader must be at depth zero, it was at:" + reader.getDepth());
+        checkArgument(reader.getType() != null,
+            "reader must be positioned at a value; call IonReader.next() first.");
```

Comment:
> call IonReader.next() first

The other common situation where `getType() == null` is the end of a stream, at which point `IonReader.next()` isn't going to help. It's much less likely that a user will call `matchCurrentValue` on an exhausted reader, though, so it may not be worth calling out. 🤷